### ISSUE TITLE
Migrating to package:characters?

### ISSUE THREAD

**Issue by domesticmouse** [petitparser/dart-petitparser#80]
Hi Lukas,

I'm curious as to your feelings about the potential for porting Petit Parser to the [Characters](https://pub.dev/packages/characters) package? 

I'm wondering if the [CharacterRange](https://pub.dev/documentation/characters/latest/characters/CharacterRange-class.html) iterator is sufficient for Petit Parser's backtracking requirements. The upside of migrating to the Character package is that Petit Parser would be parsing in terms of Unicode Grapheme clusters instead of characters.

brett

**Comment by renggli** [comment]
Hi Brett,

The [characters](https://pub.dev/packages/characters) package is great, I've already converted some of my other code to use it. For example, the [more](https://pub.dev/packages/) package uses it for its configurable [printers](https://github.com/renggli/dart-more/tree/master/lib/src/printer). I've also tried to use it for the [char_matcher](https://github.com/renggli/dart-more/tree/master/lib/src/char_matcher) in the same package (which is used in similar form in PetitParser), with less success. I found it challenging to implement character-based predicates efficiently, for example character ranges or character sets (whitespaces, letters, digits, lower-case, upper-case, ...).

PetitParser needs random access to characters. It currently does it mostly through `codeUnitAt`, however the original implementation in Smalltalk used something similar to [CharacterRange](https://pub.dev/documentation/characters/latest/characters/CharacterRange-class.html) to navigate forward and backward through the input. I expect this could easily be adopted in Dart.

What I am more concerned about is performance: I expect `moveNext`, `current` and `copy` to be the most frequently called methods. They all execute quite a bit of code, and most of them also allocate new objects. In version 2.2.0 I introduced a fast-parse mode that avoids memory allocations during parsing, which brought a huge speed improvement on some inputs for lexer-like matching. Not sure how this could be adopted with [characters](https://pub.dev/packages/characters)?

**Comment by domesticmouse** [comment]
I'm sure we'd love some feedback at https://github.com/dart-lang/characters if you have a chance =)

**Comment by Hamza5** [comment]
Now after more than a year, is there any Unicode support available? We are in 2021 and Unicode support is really required.

**Comment by renggli** [comment]
This library is designed to parse over streams of bytes (or UTF-16 characters), and as such it is agnostic to the encoding of your input. For example, the [xml](http://pub.dev/packages/xml) library is [successfully parsing](https://github.com/renggli/dart-xml/blob/4744be75ee950acbede556187180568bce373600/test/examples.dart#L213) unicode input without the need for either of the libraries to "understand" the underlying characters.

While I understand that out of the box decoding would be desirable, it comes at a hefty cost in performance. So far I haven't seen a compelling need to support it natively, so I haven't invested more time to support it. A real-world use-case where the current infrastructure doesn't work would definitely help to motivate investing time into this issue.

**Comment by RandalSchwartz** [comment]
To only slightly hijack this thread, I'm wondering if it's possible to use the higher-level logic provided by the parser to extract values out of an arbitrary jsonDecode.  It'd be nice if I could specify a pattern of a json object containing some strings and bools and a json list of further objects, and then the .map action could map that into a Dart object via a constructor.  I suspect it's only a matter of coming up with a useful set of primitives, and then using the rest of the mechanics without change, but if there's been any thought about this, I'd be interested.  I know the switch stuff makes some of this easier, but it still doesn't feel as satisfying as just writing a composable Parser rule.

**Comment by amake** [comment]
> A real-world use-case where the current infrastructure doesn't work would definitely help to motivate investing time into this issue.

I'm not sure if this fits the bill, but I just discovered that I can't make a `pattern` parser with Unicode codepoints above U+FFFF because `toCharCode` only sees the first surrogate pair.

To make everything safe for characters above U+FFFF, I think you need to replace every instance of `String.codeUnitAt` with the equivalent of `String.runes.first`.

**Comment by renggli** [comment]
As mentioned in https://github.com/petitparser/dart-petitparser/issues/147#issuecomment-1493443201 you can create a parser that accepts surrogate pairs. That said, there should probably be helpers that would make this simpler and the `pattern` parser should likely support this out of the box.

**Comment by amake** [comment]
What I want is to say `pattern('\u{20000}-\u{2FFFF}')` (this is an actual pattern I wanted to make) but this blows up.

I suppose I can make a big OptionParser with `string('\u{20000}') | string('\u20001') | ...` but this doesn't seem reasonable to me.

**Comment by renggli** [comment]
Understood, unfortunately this is currently not possible. Switching to use `runes` or `characters` everywhere comes at a hefty performance penalty for everybody, while most parsers do not require it. 

In the meantime I suggest you work around with the proposed solution, there is no need for big OptionParser:

```dart
final surrogatePair = seq2(
    pattern('\uD800-\uDBFF'),
    pattern('\uDC00-\uDFFF'),
);
final decodedSurrogatePair = surrogatePair.map2((hi, lo) =>
    0x400 * (hi.codeUnitAt(0) - 0xD800) +
    (lo.codeUnitAt(0) - 0xDC00) +
    0x10000);
final patternInQuestion = decodedSurrogatePair
    .where((value) => 0x20000 <= value && value <= 0x2FFFF);
```

**Comment by amake** [comment]
> Switching to use `runes` or `characters` everywhere comes at a hefty performance penalty for everybody, while most parsers do not require it.

I agree with this. I had two possibilities in mind:

- Only use `runes` when necessary (peek at the first code unit, and if it's a surrogate then use `runes`)
- Introduce alternate parsers that use `runes`: it is not an uncommon API design to have a `unicode` flag or even parallel set of classes/functions/etc. In this case it could be `pattern('\u{20000}-\u{2FFFF}', unicode: true)` or even `unicodePattern('\u{20000}-\u{2FFFF}')`

Do either of these seem compelling to you?

**Comment by renggli** [comment]
I like the `unicode: true` param, that matches the API of the official [RegExp](https://api.dart.dev/dart-core/RegExp/RegExp.html) class.

**Comment by amake** [comment]
I started looking at this and it's fairly hairy.

- Most parser classes make some assumption that 1 char = 1 code unit, and the conditionals and fiddling needed to handle the Unicode case make the code much harder to read.
- The top-level functions like `pattern`, `any`, etc., frequently have optional positional parameters, which prevents the addition of a `unicode` named argument. A boolean positional argument is not very nice in my opinion.

It looks to me like a parallel API would be nicer. I wouldn't blame you for not wanting to deal with that.

**Comment by renggli** [comment]
Agreed, a boolean positional argument is not desired. In this case probably better to introduce a parallel API (`patternUnicode`, `anyUnicode`, etc.) and then deprecate it again when Dart adds support to [Allow both optional positional and optional named arguments in the same function signature](https://github.com/dart-lang/language/issues/1076).

I can look into some options later today.

**Comment by renggli** [comment]
Some initial prototype of how a unicode character parser could look like: https://github.com/petitparser/dart-petitparser/compare/unicode. In the tests it seems to work well, not sure how to create a nice API yet.

**Comment by amake** [comment]
Thanks for that. I ended up with something similar but more specialized—my only use case allows me to get by with a Parser<int> because I always discard the result.
https://github.com/amake/org_parser/commit/5056bcf04b857a9a39bf0e43c5fc0066be37e413#diff-8f6acb6c09132272f801008ec6c7197d34f97ad78f3f37be4335d90618e4f5c4R20

**Comment by renggli** [comment]
I am closing this, PetitParser 7.0.0 has full support for parsing unicode characters.
